pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,233 @@
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThresholdPercentage, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.cb = cbBuilder.Build()
+	cb.metrics = newCircuitBreakerMetrics(registerer, cb.cb.State)
+	return &cb
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+	if statusCode == codes.DeadlineExceeded {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	if cb == nil {
+		return false
+	}
+	return cb.active.Load()
+}
+
+func (cb *circuitBreaker) setActive() {
+	if cb == nil {
+		return
+	}
+	cb.active.Store(true)
+}
+
+// tryAcquirePermit tries to acquire a permit to use the circuit breaker and returns whether a permit was acquired.
+// If it was possible to acquire a permit, a status true and no error are returned. The acquired permit must be
+// returned by a call to finishPushRequest.
+// If it was not possible to acquire a permit, a status false is returned. In this case no call to finishPushRequest
+// is needed. If the permit was not acquired because of an error, that causing error is returned as well.
+func (cb *circuitBreaker) tryAcquirePermit() (bool, error) {
+	if !cb.isActive() {
+		return false, nil
+	}
+	if !cb.cb.TryAcquirePermit() {
+		cb.metrics.circuitBreakerResults.WithLabelValues(resultOpen).Inc()
+		return false, newCircuitBreakerOpenError(cb.cb.RemainingDelay())
+	}
+	return true, nil
+}
+
+func (cb *circuitBreaker) recordResult(err error) {
+	if !cb.isActive() {
+		return
+	}
+	if err != nil && isFailure(err) {
+		cb.cb.RecordFailure()
+		cb.metrics.circuitBreakerResults.WithLabelValues(resultError).Inc()
+	} else {
+		cb.metrics.circuitBreakerResults.WithLabelValues(resultSuccess).Inc()
+		cb.cb.RecordSuccess()
+	}
+}
+
+// finishPushRequest should be called to complete the push request executed upon a
+// successfully acquired circuit breaker permit.
+// It records the result of the push request with the circuit breaker. Push requests
+// that lasted longer than the configured timeout are treated as a failure.
+// The returned error is only used for testing purposes.
+func (cb *circuitBreaker) finishPushRequest(ctx context.Context, duration time.Duration, pushErr error) error {
+	if !cb.isActive() {
+		return nil
+	}
+	if cb.cfg.testModeEnabled {
+		if testDelay, ok := ctx.Value(testDelayKey).(time.Duration); ok {
```

Comment:
Since we're explicitly passing duration to this function, do we still need artificial `testDelayKey`?

Reply:
We actually do, because the function we call in the relevant test is `PushToStorage()` and we have no other mechanism of letting the `PushToStorage()` know that its request should last longer.

Reply:
I have actually found another way to do this, so I have removed the `test-delay` key from the context.